Replaced a typo with rockchip_pmu_set_idle_request

Miouyouyou opened the pull request:
Thanks to DuckDuckGo for the suggestion when I was searching the
mistyped symbol.

Well, I think it's this one. It almost has the same name AND the same
signature. However, I currently do not understand what the code does so
I might be wrong.
The module compiles and loads fine with this correction. However, I do
not know if it runs fine.

This *still* needs a patched mainline kernel in order that exports this symbol.
The appropriate patch is available here:
https://github.com/Miouyouyou/MyyQi/blob/a1e040fb36bf4b34a926464ef062c1bd4fe7faaf/patches/kernel/v4.10-rc4/0013-Export-rockchip_pmu_set_idle_request-for-out-of-tree.patch

Signed-off-by: Myy <myy@miouyouyou.fr>

phhusson commented:
*facepalm*
I was actually using this: (cherry-picked from rk 4.4)
https://github.com/phhusson/linux/commit/9f1cd9cf07bb08e65ee5a12f7f6231cf59eb708b

So, thanks for that :)

As for whether this runs, my current status is that it almost runs on files slow enough (i.e. h265, or some h264 when I have stdout with lots of logs on ssh through WiFi) using gstreamer-rockchip packages.

There are two instances of rockchip_iovmm_invalidate_tlb that are currently commented out, this is the most probable reason of why it doesn't always succeed.
I need to find another way to clean TLB, but ATM I haven't been sucessful.

I strongly recommend you to go to #linux-rockchip @ irc.freenode.net. I'm there as well as some rockchip kernel maintainer and some engineers working for rockchip.

Miouyouyou committed:
Replaced a typo with rockchip_pmu_set_idle_request

Thanks to DuckDuckGo for the suggestion when I was searching the
mistyped symbol.

Well, I think it's this one. It almost has the same name AND the same
signature. However, I currently do not understand what the code does so
I might be wrong.
The module compiles and loads fine with this correction. However, I do
not know if it runs fine.

Signed-off-by: Myy <myy@miouyouyou.fr>

Miouyouyou commented:
I wonder how they got their thing working... After checking this VPU code, it seems that there are *at least* 3 different implementations of it... :sweat_smile:  This one, google chrome one and rockchip-linux one all differ in some points.

I'm glad that @mac_l1 provided the address of yours, as [trying to port rockchip-linux/kernel vpu code](https://gitlab.com/Miouyouyou/MyyQi/snippets/35340) led me to nowhere... :sweat: 

I'll try to see if I can get better performances by identifying the differences between kernel 4.4 VPU code and this one, and also testing it on a Debian RK3288 board (MiQi) with pre-patched packages installed. 

Still, cleaning TLB seems to be an expensive operation. I don't know how VPU code works though, when watching a movie. Does it decode in real-time or does it buffer everything ? 

I'll try to pass more often on #rockchip-linux, though ! With the same handle, 'Myy'.

phhusson commented:
The one I have here comes directly from rockchip-linux/kernel. Though there has been a revamp quite recently (november I'd say), this code comes after the revamp, and yours seems to be before it.

chromium's VPU (which was described in this very git before) is v4l2-based, so it was thought to be more likely to be merged by mainline, but this didn't happen, because of userland API issues. Also it can only decode h264 and vp8.

The driver currently developed by Rockchip for GNU/Linux use is the one here, to be used with https://github.com/rockchip-linux/gstreamer-rockchip and libmpp, and it can decode h264, h265 and vp8. (mpp is capable of many more codecs and of encoding)
They hope to have a solution for mainline, but nothing short-term.

As for TLB (fwiw it's iommu's tlb, not cpu's tlb), it doesn't happen often, I'd say few times per frame tops, so it should be ok performance wise.

phhusson commented:
BTW, I have the proper history of this vcodec driver there: https://github.com/phhusson/linux/tree/vcodec-mod

Miouyouyou commented:
Ah ! That's interesting. I might be able to compile gstreamer-rockchip easily on my Gentoo board. Libmpp compiled nicely, last time I tried, but I didn't have working drivers so the testing procedures went badly. :stuck_out_tongue: 

Well, I'll try to create two branches, one with the comments on the IOMMU TLB code and one without, and let some video decoding software fans do the benchmarks themselves.

Nice to have the full history of this codec, as the chromium one and the rockchip-kernel one didn't seem to connect (notably on the IOMMU code. I wasn't able to identify when the IOMMU code were added and when they decided to try using the [Common IOMMU code branch that was never ported](https://lkml.org/lkml/2016/2/19/79)).
I'll bookmark this link, in case something goes wrong with the current VPU code.

The UAPI changes of the chromium were huge and changed a lot of data structures by putting new fields in-between others, which break a lot of software compiled with untouched headers, so I can understand they could not merge it.

phhusson commented:
I have to warn again, this driver doesn't really work. 
I recommend getting a 1080p 10bit HEVC file to have higher chances of success.

I'm pretty sure the TLB thingy is really needed.
